### PR TITLE
Added Simple Geometry Containers and Readers (MAP-59/60)

### PR DESCRIPTION
**Pull request by wchen329**
Added some containers for simple geometry and tests for them. Also added a factory for Simple Geometry.

### CODE REVIEW

File: cdm/src/main/java/ucar/nc2/ft2/coverage/adapter/SimpleGeometryCS.java [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+  // LOOK another possibility is a scalar runtime and a 1D time offset
+
+  @Override
+  public CoordinateAxis1DTime getTimeAxis() {
+    return (CoordinateAxis1DTime) super.getTimeAxis();
+  }
+
+  /*
+   @Override
+ public List<CalendarDate> getCalendarDates() {
```
Comment (skaymen) [inline cdm/src/main/java/ucar/nc2/ft2/coverage/adapter/SimpleGeometryCS.java]: Should this code still be included if commented out?

File: cdm/src/test/java/ucar/nc2/ft/TestCFSimpleGeometryCont.java [hunk 2]
```diff
@@ -0,0 +1,316 @@
+package ucar.nc2.ft2.coverage.simpgeometry;
```
Comment (skaymen) [inline cdm/src/test/java/ucar/nc2/ft/TestCFSimpleGeometryCont.java]: Should these tests be in separate classes?

**Review by skaymen — APPROVED** [review]
Everything seems reasonable to me. A couple style comments, and test data sets could be made smaller if necessary, but no real problems. Nice work on this.

File: cdm/src/main/java/ucar/nc2/ft2/coverage/adapter/SimpleGeometryCS.java (hunk 1 above)
Comment (wchen329) [reply cdm/src/main/java/ucar/nc2/ft2/coverage/adapter/SimpleGeometryCS.java]: This was forked from the original as well, could probably delete it.

File: cdm/src/test/java/ucar/nc2/ft/TestCFSimpleGeometryCont.java (hunk 2 above)
Comment (wchen329) [reply cdm/src/test/java/ucar/nc2/ft/TestCFSimpleGeometryCont.java]: Perhaps but most tests are in ucar.nc2! Could revisit later.

File: cdm/src/main/java/ucar/nc2/dataset/conv/CF1Convention.java [hunk 3]
```diff
@@ -171,42 +175,69 @@ public void augmentDataset(NetcdfDataset ds, CancelTask cancelTask) throws IOExc
...
       if (v.findAttribute(CF.GEOMETRY) != null) {
-      	v.addAttribute(new Attribute(CF.NODES, ds.findAttValueIgnoreCase(v, CF.NODES, "")));
-      	v.addAttribute(new Attribute(CF.NODE_COUNT, ds.findAttValueIgnoreCase(v, CF.NODE_COUNT, "")));
-      	v.addAttribute(new Attribute(CF.NODE_COORDINATES, ds.findAttValueIgnoreCase(v, CF.NODE_COORDINATES, "")));
-      	v.addAttribute(new Attribute(CF.PART_NODE_COUNT, ds.findAttValueIgnoreCase(v, CF.PART_NODE_COUNT, "")));
-      	if ("polygon".equalsIgnoreCase(ds.findAttValueIgnoreCase(v, CF.GEOMETRY_TYPE, ""))) {
-      		v.addAttribute(new Attribute(CF.INTERIOR_RING, ds.findAttValueIgnoreCase(v, CF.INTERIOR_RING, "")));
+    	  
+    	Attribute container = v.findAttribute(CF.GEOMETRY);
+    	Variable coordsvar = ds.findVariable(container.getStringValue());
```
Comment (cschroedl-gov) [inline cdm/src/main/java/ucar/nc2/dataset/conv/CF1Convention.java:186]: Good call to dynamically look up the geometry container. I believe that is consistent with the convention and should therefore permit us to read in files that have multiple data variables that reference simple geometries. Good stuff!

File: cdm/src/main/java/ucar/nc2/ft2/coverage/adapter/DtCoverageCSBuilder.java [hunk 4]
```diff
@@ -62,6 +68,7 @@ public static String describe(NetcdfDataset ds, CoordinateSystem cs, Formatter e
   List<CoordinateAxis> otherAxes;
   List<CoordinateAxis> allAxes;
   List<CoordinateTransform> coordTransforms;
+  SimpleGeometryReader geometry_reader;
```
Comment (cschroedl-gov) [inline cdm/src/main/java/ucar/nc2/ft2/coverage/adapter/DtCoverageCSBuilder.java]: It looks like other instance variables in this class use camelCasing. Is there a reason this instance variable should use a different naming convention?

File: cdm/src/main/java/ucar/nc2/constants/AxisType.java [hunk 5]
```diff
@@ -28,8 +28,11 @@
   RadialElevation(6),
   Spectral(1),
   TimeOffset(1,"TO"),  // time offset from runtime / reference time
-  Dimension(99,"Dim");      // used for dimension axis (experimental);
-
+  Dimension(99,"Dim"),      // used for dimension axis (experimental);
```
Comment (cschroedl-gov) [inline cdm/src/main/java/ucar/nc2/constants/AxisType.java:31]: The CF Convention docs describe something called a "Discrete Axis". Is that at all relevant here?
https://github.com/cf-convention/cf-conventions/blob/master/ch04.adoc#discrete-axis

It was present in CF [even in version 1.6](http://cfconventions.org/cf-conventions/v1.6.0/cf-conventions.html#discrete-axis), and I would guess it would have been added as an AxisType already if it were relevant, but thought I would check.

File: cdm/src/main/java/ucar/nc2/ft2/coverage/adapter/DtCoverageCSBuilder.java [hunk 6]
```diff
@@ -264,6 +283,102 @@ private FeatureType classify () {
...
+  /**
+   * Given a certain Polygon variable name and geometry begin and end indicies, returns a list of Simple Geometry Polygon
+   * 
+   * @param name
+   * @param index_begin
+   * @param index_end
+   * @return
+   */
+  public List<Polygon> getPolygons(String name, int index_begin, int index_end) {
+	  List<Polygon> poly_list = new ArrayList<Polygon>();
```
Comment (cschroedl-gov) [inline cdm/src/main/java/ucar/nc2/ft2/coverage/adapter/DtCoverageCSBuilder.java]: Similar question on naming convention here. Most of the other local variables and parameters in this class use camelCase. Is there a reason why the local variables and method parameters in the added code should use a different naming convention?

File: cdm/src/main/java/ucar/nc2/ft2/coverage/adapter/SimpleGeometryCS.java [hunk 7]
```diff
@@ -0,0 +1,121 @@
...
+  
+  // LOOK another possibility is a scalar runtime and a 1D time offset
+  
+  @Override
+  public CoordinateAxis1DTime getTimeAxis() {
+    return (CoordinateAxis1DTime) super.getTimeAxis();
+  }
+
+  @Override
+  public CalendarDateRange getCalendarDateRange() {
```
Comment (cschroedl-gov) [inline cdm/src/main/java/ucar/nc2/ft2/coverage/adapter/SimpleGeometryCS.java]: A few of these methods look substantially similar to the implementations in the superclass. In this case, 

https://github.com/usgs/thredds/blob/5.0.0-usgs/cdm/src/main/java/ucar/nc2/ft2/coverage/adapter/DtCoverageCS.java#L395

Would it be relevant to simplify this subclass' implementation by refraining from defining the methods in the subclass? Alternatively, if a method must be `@Override`n, could the subclass' implementation be simplified by calling the super class' implementation of the same method?

File: cdm/src/main/java/ucar/nc2/ft2/coverage/simpgeometry/SimpleGeometryKitten.java [hunk 8]
```diff
@@ -0,0 +1,89 @@
...
+ * 
+ * A Simple Geometry Kitten can go find the beginning and end indicies of
+ * a simple geometry X and Y within a variable. But first the Kitten needs a few variables to find it in.
+ * The kitten remembers previous adventures so if the kitten is tasked to find a Simple Geometry close to one before it
+ * will find it faster.
+ * 
+ * @author wchen@usgs.gov
+ *
+ */
+public class SimpleGeometryKitten {
```
Comment (cschroedl-gov) [inline cdm/src/main/java/ucar/nc2/ft2/coverage/simpgeometry/SimpleGeometryKitten.java]: I think this name is cute, but I can't tell, is the Kitten a performance optimization, or a necessity? If it is a performance optimization, can it be removed? If it is a necessity, can it be given a name that makes its purpose more obvious?

Comment (cschroedl-gov) [reply cdm/src/main/java/ucar/nc2/ft2/coverage/simpgeometry/SimpleGeometryKitten.java]: Whether or not it applies here, it's important to know about the following wisdom from the CS guru Donald Knuth, or as I like to call him, Donald "How is your name on so many important papers?" Knuth.

> Programmers waste enormous amounts of time thinking about, or worrying about, the speed of noncritical parts of their programs, and these attempts at efficiency actually have a strong negative impact when debugging and maintenance are considered. We should forget about small efficiencies, say about 97% of the time: premature optimization is the root of all evil. Yet we should not pass up our opportunities in that critical 3%

I'm not as dramatic as Knuth, so I wouldn't call anything "evil". I am wondering if we can do without this in the short term.

File: cdm/src/test/java/ucar/nc2/ft/simpgeometry/TestCFSimpleGeometryCont.java [hunk 9]
```diff
@@ -0,0 +1,319 @@
...
+import java.util.List;
+
+/**
+ * Very simple tests for Simple Geometry container objects.
+ * 
+ * @author wchen@usgs.gov
+ *
+ */
+public class TestCFSimpleGeometryCont {
+
```
Comment (cschroedl-gov) [inline cdm/src/test/java/ucar/nc2/ft/simpgeometry/TestCFSimpleGeometryCont.java]: Most of the logic in the implementation and the test approximates the logic of the native LinkedList class. Is there a way to simplify the implementation and the test by using the native LinkedList class instead?

File: cdm/src/test/java/ucar/nc2/ft/simpgeometry/TestCFSimpleGeometryCont.java [hunk 10]
```diff
@@ -0,0 +1,319 @@
...
+/**
+ * Very simple tests for Simple Geometry container objects.
+ * 
+ * @author wchen@usgs.gov
+ *
+ */
+public class TestCFSimpleGeometryCont {
+	
+	private static final double delt = 0.00;
+	private static final int testsize = 100000;
```
Comment (cschroedl-gov) [inline cdm/src/test/java/ucar/nc2/ft/simpgeometry/TestCFSimpleGeometryCont.java]: How is test 100 different from test 101?
What is the smallest possible test size that addresses corner cases?

**Review by cschroedl-gov — CHANGES_REQUESTED** [review]

File: cdm/src/main/java/ucar/nc2/ft2/coverage/adapter/DtCoverageCSBuilder.java (hunk 4 above)
Comment (wchen329) [reply cdm/src/main/java/ucar/nc2/ft2/coverage/adapter/DtCoverageCSBuilder.java]: None in particular, I will change it so it is consistent, thanks!

File: cdm/src/main/java/ucar/nc2/ft2/coverage/adapter/DtCoverageCSBuilder.java (hunk 6 above)
Comment (wchen329) [reply cdm/src/main/java/ucar/nc2/ft2/coverage/adapter/DtCoverageCSBuilder.java]: Ditto:

None in particular, I will change it so it is consistent, thanks!

File: cdm/src/test/java/ucar/nc2/ft/simpgeometry/TestCFSimpleGeometryCont.java (hunk 9 above)
Comment (wchen329) [reply cdm/src/test/java/ucar/nc2/ft/simpgeometry/TestCFSimpleGeometryCont.java]: The motivation for using a Polygon class with next and previous to represent multipolygons was so that they could be treated alike and hence be placed in the same containers since they are very similar (a multipolygon is basically the same except with a "next" while a regular polygon has null "next").

File: cdm/src/main/java/ucar/nc2/ft2/coverage/simpgeometry/SimpleGeometryKitten.java (hunk 8 above)
Comment (wchen329) [reply cdm/src/main/java/ucar/nc2/ft2/coverage/simpgeometry/SimpleGeometryKitten.java]: Sure sounds good. I'll change it to "SimpleGeometryIndexer" which is more descriptive

Comment (wchen329) [reply cdm/src/main/java/ucar/nc2/ft2/coverage/simpgeometry/SimpleGeometryKitten.java]: Though I will remark that it was a reference to how "Dogs" are used as search engine mascots but instead this is a cat.

File: cdm/src/test/java/ucar/nc2/ft/simpgeometry/TestCFSimpleGeometryCont.java (hunk 10 above)
Comment (wchen329) [reply cdm/src/test/java/ucar/nc2/ft/simpgeometry/TestCFSimpleGeometryCont.java]: Yeah I'll address corner cases, thanks! And split them up.

File: cdm/src/main/java/ucar/nc2/ft2/coverage/adapter/SimpleGeometryCS.java (hunk 7 above)
Comment (wchen329) [reply cdm/src/main/java/ucar/nc2/ft2/coverage/adapter/SimpleGeometryCS.java]: Seems like these are from the GridCS which SimpleGeometryCS was originally forked from. These may not be relevant to begin with, so they might require cleanup. Thanks!

File: cdm/src/main/java/ucar/nc2/ft2/coverage/simpgeometry/SimpleGeometryKitten.java (hunk 8 above)
Comment (wchen329) [reply cdm/src/main/java/ucar/nc2/ft2/coverage/simpgeometry/SimpleGeometryKitten.java]: Or actually "SimpleGeometryIndexFinder"

File: cdm/src/main/java/ucar/nc2/constants/AxisType.java (hunk 5 above)
Comment (wchen329) [reply cdm/src/main/java/ucar/nc2/constants/AxisType.java:31]: Something to keep in mind for sure, but it is not needed as of now, thanks though!

File: cdm/src/main/java/ucar/nc2/ft2/coverage/simpgeometry/SimpleGeometryKitten.java (hunk 8 above)
Comment (wchen329) [reply cdm/src/main/java/ucar/nc2/ft2/coverage/simpgeometry/SimpleGeometryKitten.java]: Concerning the premature optimization thing, the way it is implemented in (formerly what was) the Kitty is, is necessary for the common case, which is taking in several sequential geometries. Having it would seem to have a decent impact on performance here.

File: cdm/src/test/java/ucar/nc2/ft/TestCFSimpleGeometryCont.java (hunk 2 above)
Comment (wchen329) [reply cdm/src/test/java/ucar/nc2/ft/TestCFSimpleGeometryCont.java]: They will be split in the end to be in agreement with convention, thanks!

File: cdm/src/main/java/ucar/nc2/constants/AxisType.java (hunk 5 above)
Comment (cschroedl-gov) [reply cdm/src/main/java/ucar/nc2/constants/AxisType.java:31]: 👍

File: cdm/src/main/java/ucar/nc2/ft2/coverage/simpgeometry/CFPoint.java [hunk 11]
```diff
@@ -0,0 +1,256 @@
...
+	 * 
+	 * Assumes that data is null.
+	 * 
+	 * @param x - the x coordinate of the point
+	 * @param y - the y coordinate of the point
+	 * @param prev - previous point if part of a multipoint
+	 * @param next - next point if part of a multipoint
+	 */
+	public CFPoint(double x, double y, CFPoint prev, CFPoint next) {
+		this.next = next;
```
Comment (cschroedl-gov) [inline cdm/src/main/java/ucar/nc2/ft2/coverage/simpgeometry/CFPoint.java]: It looks like this constructor has a lot in common with the constructor that permits a caller to specify the data they want to be associated with the `CFPoint`. Could the body of this constructor be simplified to something like:
```
this(x, y, prev, next, null);
```
...and still accomplish the same thing?

File: cdm/src/main/java/ucar/nc2/ft2/coverage/adapter/DtCoverageCSBuilder.java (hunk 6 above)
Comment (cschroedl-gov) [reply cdm/src/main/java/ucar/nc2/ft2/coverage/adapter/DtCoverageCSBuilder.java]: 👍

File: cdm/src/main/java/ucar/nc2/ft2/coverage/adapter/SimpleGeometryCS.java (hunk 7 above)
Comment (cschroedl-gov) [reply cdm/src/main/java/ucar/nc2/ft2/coverage/adapter/SimpleGeometryCS.java]: 👍

File: cdm/src/main/java/ucar/nc2/ft2/coverage/simpgeometry/SimpleGeometryKitten.java (hunk 8 above)
Comment (cschroedl-gov) [reply cdm/src/main/java/ucar/nc2/ft2/coverage/simpgeometry/SimpleGeometryKitten.java]: Would the absence of this optimization cause a common use case to take more than 5 minutes to execute?

File: cdm/src/main/java/ucar/nc2/ft2/coverage/simpgeometry/CFLine.java [hunk 12]
```diff
@@ -0,0 +1,273 @@
...
+				}
+				
+				//Clean up
+				tail = tail.getPrev();
+				if(tail != null) tail.setNext(null);
+			}
+		}
+		
+		catch (IOException e) {
+			e.printStackTrace();
```
Comment (cschroedl-gov) [inline cdm/src/main/java/ucar/nc2/ft2/coverage/simpgeometry/CFLine.java:239]: `exception.printStackTrace()` is a handy debugging tool on simpler projects that don't rely on logging tools. On big projects like THREDDS, the quantity of log messages gets so overwhelming that formal logging tools are needed to address the flood of information.  This permits people inspecting the logs and configuring the app to choose what level of detail they want, and to send different kinds of information to different log files. Would you be able to help users of THREDDS better scope this troubleshooting infromation by logging these Exceptions instead of using `e.printStackTrace()`?

DZone has an article on this:
https://dzone.com/articles/java-top-5-exception-handling

I think THREDDS uses SLF4J. I would encourage you to check out other THREDDS classes usages of this logging framework.

File: cdm/src/test/java/ucar/nc2/ft2/coverage/simpgeometry/TestReadPoint.java [hunk 13]
```diff
@@ -0,0 +1,47 @@
...
+public class TestReadPoint {
+
+	private SimpleGeometryReader newReader()
+	{
+		String filepath = TestDir.cdmLocalTestDataDir + "dataset/SimpleGeos/avg_temp_3gage_5timesteps.nc";
+		NetcdfDataset dataset = null;
+		
+		try {
+		
+			dataset = NetcdfDataset.openDataset(filepath);
```
Comment (cschroedl-gov) [inline cdm/src/test/java/ucar/nc2/ft2/coverage/simpgeometry/TestReadPoint.java]: Would you be able to add an assertion to this test? Maybe just a not null assertion to start out?

File: cdm/src/test/java/ucar/nc2/ft2/coverage/simpgeometry/TestReadPolygon.java [hunk 14]
```diff
@@ -0,0 +1,64 @@
...
+ *
+ */
+public class TestReadPolygon {
+
+	/**
+	 * Tests ability to read and construct Polygons and Multipolygons.
+	 * 
+	 */
+	@Test
+	public void testMultiPolygonsPolygons() {
```
Comment (cschroedl-gov) [inline cdm/src/test/java/ucar/nc2/ft2/coverage/simpgeometry/TestReadPolygon.java]: I like this test method!

**Comment by cschroedl-gov** [comment]
Thanks for the improvements @wchen329 !

A few general things:
 * There are several local variables and parameters that are not using camelCasing. Is there widespread precedent for using camelCasing or underscoring in the THREDDS codebase? Could the names be updated to use the most prominent convention in THREDDS?
 * Do the current implementations of CFPoint, CFLine, CFPolygon abide by the Single Responsibility Principle? https://en.wikipedia.org/wiki/Single_responsibility_principle . A common way to separate concerns in Java projects is to have a few simple/dumb beans represent the business objects, and a few other classes that understand how to construct those beans from different data sources. For example, a Receipt bean could know about a bunch of items and their prices and there could be multiple readers that know how to read it into Memory from various file formats (JSON, XML, CSV). With this in mind, what do you think about moving the `setup` methods to new small CF(Point/Line/Polygon)Reader classes? Are there other ways that the Simple Geometry entities could be effectively separated from how they are loaded from data sources?

File: cdm/src/main/java/ucar/nc2/ft2/coverage/simpgeometry/CFLine.java (hunk 12 above)
Comment (wchen329) [reply cdm/src/main/java/ucar/nc2/ft2/coverage/simpgeometry/CFLine.java:239]: I was aware of SLF4J being used in THREDDS, however, it looks like there's some spottiness in it being included in NetCDF-All. Will look into it, thanks!

File: cdm/src/main/java/ucar/nc2/ft2/coverage/simpgeometry/CFPoint.java (hunk 11 above)
Comment (wchen329) [reply cdm/src/main/java/ucar/nc2/ft2/coverage/simpgeometry/CFPoint.java]: Overloading that constructor was done to allow the choice to set data on construction since Java doesn't seem to have a way (1) to set default values unlike better languages. It's probably part of the "enforced explicitness" of Java since Java has many examples of enforcing lack of elegance for the sake of simplicity.

(1) Might be wrong about that but such a design (the disallowance of default values) seems very in line with the way Java is designed.

File: cdm/src/test/java/ucar/nc2/ft2/coverage/simpgeometry/TestReadPolygon.java (hunk 14 above)
Comment (wchen329) [reply cdm/src/test/java/ucar/nc2/ft2/coverage/simpgeometry/TestReadPolygon.java]: Thanks! Perhaps this will be a model for the similar Read Point and Read Line as well.

File: cdm/src/main/java/ucar/nc2/ft2/coverage/simpgeometry/CFPoint.java (hunk 11 above)
Comment (wchen329) [reply cdm/src/main/java/ucar/nc2/ft2/coverage/simpgeometry/CFPoint.java]: In any case, we could delete the constructor since the data-set-able one can do the same thing. At the same time though it will make the data-set-able one just more obtuse. Thanks!

File: cdm/src/test/java/ucar/nc2/ft2/coverage/simpgeometry/TestReadPoint.java (hunk 13 above)
Comment (wchen329) [reply cdm/src/test/java/ucar/nc2/ft2/coverage/simpgeometry/TestReadPoint.java]: Oh yeah this test is not finished, still need working Point data and Line data. Will get both the Line and Point tests finished once those are fixed.

File: cdm/src/main/java/ucar/nc2/dataset/conv/CF1Convention.java (hunk 3 above)
Comment (wchen329) [reply cdm/src/main/java/ucar/nc2/dataset/conv/CF1Convention.java:186]: Awesome! Might add some more heuristics though for finding geometry_container in dataless geometry sets, though it might not be needed.

File: cdm/src/test/java/ucar/nc2/ft2/coverage/simpgeometry/TestReadPoint.java (hunk 13 above)
Comment (wchen329) [reply cdm/src/test/java/ucar/nc2/ft2/coverage/simpgeometry/TestReadPoint.java]: I will add a Not Null Check though on the Reader, thanks!

**Comment by wchen329** [comment]
Thanks for all the feedback Carl!

CamelCasing is definitely that which is most widespread through THREDDS, I can update all the casing (I've placed at least) though it will take some time.

At this level you're not dealing with XMLs and NetCDF stuff and etcetera- reading from CDM compliant files happens at the NetCDF File layer so it's already done by the time "SimpleGeometryReader" even begins to exist. What the reader deals with is constructing these Polygons from a loaded NetCDFFile/NetCDFDataset which holds Simple Geometry information.

One point of concern may be indeed different standards- as it suggests CFPoint/Line/Polygon are very specific to the CF standard (as probably hinted by the name). This was the point of the interface Point, Line and Polygon and the idea is that the convention used would be passed in and based on the convention the SimpleGeometryReader would use the right type of Polygon/Line/Point to construct (which implements Polygon or a simple geometry interface) and send it back (Polygon, Line and Point should all have this "setup" function which varies on standard implementation). Unfortunately the interfaces are not as generic as needed [in the interest of saving time to get some minimally viable product out of the door] and for now are only placeholders (which only are guaranteed to work with the CF Convention).

It might be "a thing" to separate those types of things or whatever in Java, but I don't see any real advantage here. We could move them to a new class but keep in mind I will be the one writing the tests (and a new set of even more classes mean even more tests and the current reader tests aren't even finished).

File: cdm/src/main/java/ucar/nc2/ft2/coverage/simpgeometry/CFPoint.java (hunk 11 above)
Comment (cschroedl-gov) [reply cdm/src/main/java/ucar/nc2/ft2/coverage/simpgeometry/CFPoint.java]: > Overloading that constructor was done to allow the choice to set data on construction

Yep! Makes sense!

 > we could delete the constructor since the data-set-able one can do the same thing. At the same time though it will make the data-set-able one just more obtuse.

Sorry, I don't think I communicated well on the first comment. My curiosity centers on whether we can deduplicate code. Deleting the constructor that does not permit the caller to set the data set is one solution to deduplicating code. I agree with your assessment that deleting that constructor might make the data-set-able one a bit more obtuse.

Your observation about Java's lack of default argument values is astute and pertinent. Default argument values are closely related to optional parameters. In Java optional constructor parameters are often implemented by creating constructors with shorter argument lists that call a constructor with longer argument lists.  In Java constructors can call other constructors via `this(...)`.

There's a [concise example of this on Stack Overflow](https://stackoverflow.com/a/7428077).

```
public Foo(int param1, int param2)
{
    this.param1 = param1;
    this.param2 = param2;
}

public Foo(int param1)
{
    this(param1, 2);
}
```

This technique (sometimes called "telescoping constructors") does not scale well to constructors that have a ton of parameters (like 10 or more), but it is pretty commonplace for constructors with fewer parameters.

All that said... just because something is commonplace doesn't mean that it applies everywhere. Hopefully this follow-up helps better illuminate what I referenced in the first comment so that you can use your judgement about whether it applies here.

File: cdm/src/main/java/ucar/nc2/dataset/conv/CF1Convention.java (hunk 3 above)
Comment (cschroedl-gov) [reply cdm/src/main/java/ucar/nc2/dataset/conv/CF1Convention.java:186]: 👍

File: cdm/src/main/java/ucar/nc2/ft2/coverage/simpgeometry/CFLine.java (hunk 12 above)
Comment (wchen329) [reply cdm/src/main/java/ucar/nc2/ft2/coverage/simpgeometry/CFLine.java:239]: A new ticket shall be made for that!

**Review by cschroedl-gov — APPROVED** [review]

**Review by cschroedl-gov — APPROVED** [review]
